File: packages/client-journal/src/masking.test.ts
import { readFileSync } from 'node:fs';
import path from 'node:path';
import process from 'node:process';
import { describe, expect, it } from 'vitest';
import { MASK, MASK_INPUT_LIMIT, barePath, masked } from './masking';

interface Vectors {
  readonly mask: string;
  readonly masked: readonly { readonly input: string; readonly output: string }[];
  readonly paths: readonly { readonly url: string; readonly path: string }[];
}

const VECTORS = JSON.parse(readFileSync(path.resolve(process.cwd(), 'masking-vectors.json'), 'utf8')) as Vectors;

// REQ-CLIENT-JOURNAL-005
describe('общие примеры маскирования', () => {
  it('знак маски совпадает с общим', () => {
    expect(MASK).toBe(VECTORS.mask);
  });

  it('маскирование даёт записанный результат', () => {
    for (const { input, output } of VECTORS.masked) expect(masked(input), input).toBe(output);
  });

  it('от адреса остаётся путь без источника, строки запроса и фрагмента', () => {
    for (const { url, path: bare } of VECTORS.paths) expect(barePath(url), url).toBe(bare);
  });
});

// REQ-CLIENT-JOURNAL-004
describe('работа маскирования ограничена', () => {
  it('текст длиннее предела обрезается до маскирования', () => {
    const long = `${'a '.repeat(MASK_INPUT_LIMIT)}ivan@example.org`;
    const result = masked(long);
    expect(result.length).toBe(MASK_INPUT_LIMIT);
    expect(result).not.toContain('@');
  });
});

File: platform-support/src/test/java/io/github/apocarteres/platform/support/internal/MaskingTest.java
package io.github.apocarteres.platform.support.internal;

import static org.assertj.core.api.Assertions.assertThat;

import java.nio.file.Path;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import tools.jackson.databind.JsonNode;
import tools.jackson.databind.json.JsonMapper;

// REQ-SUPPORT-006, REQ-CLIENT-JOURNAL-005
class MaskingTest {

  private final JsonNode vectors = JsonMapper.builder().build()
    .readTree(Path.of("..", "packages", "client-journal", "masking-vectors.json").toFile());

  @Test
  @DisplayName("Сервер маскирует по тем же общим примерам, что и журнал клиента")
  void masksLikeTheClient() {
    assertThat(Masking.MASK).isEqualTo(vectors.get("mask").asString());
    for (JsonNode vector : vectors.get("masked")) {
      assertThat(Masking.masked(vector.get("input").asString())).as(vector.get("input").asString())
        .isEqualTo(vector.get("output").asString());
    }
  }

  @Test
  @DisplayName("Сервер выделяет путь по тем же общим примерам, что и журнал клиента")
  void cutsPathsLikeTheClient() {
    for (JsonNode vector : vectors.get("paths")) {
      assertThat(Masking.barePath(vector.get("url").asString())).as(vector.get("url").asString())
        .isEqualTo(vector.get("path").asString());
    }
  }

  @Test
  @DisplayName("Текст длиннее предела обрезается до маскирования")
  void inputIsBounded() {
    String masked = Masking.masked("a ".repeat(Masking.INPUT_LIMIT) + "ivan@example.org");
    assertThat(masked).hasSize(Masking.INPUT_LIMIT).doesNotContain("@");
  }
}
